.github/workflows/changelog-checker.yml

```diff
@@ -0,0 +1,47 @@
+# This workflow checks that there is either a 'pr/no-changelog' label applied to a PR
+# or there is a changelog/<pr number>.txt file associated with a PR for a changelog entry
+
+name: Check Changelog
+
+on:
+  pull_request:
+    types: [opened, synchronize, labeled, unlabeled]
+    # Runs on PRs to master
+    branches:
+      - master/*
+
+jobs:
+  # checks that a changelog entry is present for a PR
+  changelog-check:
+    # If there  a `pr/no-changelog` label we ignore this check
+    if: "!contains(github.event.pull_request.labels.*.name, 'pr/no-changelog')"
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          ref: ${{ github.event.pull_request.head.sha }}
+          fetch-depth: 0 # by default the checkout action doesn't checkout all branches
+      - name: Check for changelog entry in diff
+        run: |
+          # check if there is a diff in the changelog directory
+          changelog_files=$(git --no-pager diff --name-only HEAD "$(git merge-base HEAD "origin/${{ github.event.pull_request.base.ref }}")" -- changelog/${{ github.event.pull_request.number }}.txt)
```

Comment:
Might be clearer to break this up into multiple steps.

Reply:
This part I've copied from Consul's version (https://github.com/hashicorp/consul/pull/9781/files), and I'm inclined to keep it identical although I agree it's not very readable.

Reply:
How does this work for changelog file(s) that is referencing an external PR, e.g. `_12345.txt`?

Reply:
Will this break on backport PRs to the `release/*` branches since we are cherry-picking and not renaming this file?

Reply:
> How does this work for changelog file(s) that is referencing an external PR, e.g. `_12345.txt`?

The check will fail. The check isn't required, and can be skipped by adding a label, if you don't want a failing check when you merge. This check is intended as a gentle nudge: "hey, did you want a changelog for this?", so hopefully it would prompt someone to verify that their _ changelog file is present and formatted correctly. I have updated the error message though.

Based on this comment I took a look at some options for handling more variations on changelog entries - with an _, maybe with multiple files, maybe you're adding a changelog file to a PR where the number doesn't match for some other reason, etc. We could certainly handle more cases with this check, but there are a lot of possible edge cases, and I think it might be best to just have this handle the simple case of: I have a PR, and if it should have a changelog entry, the changelog entry should probably match the PR number.

> Will checking the PR number break on backport PRs to the release/* branches since we are cherry-picking and not renaming this file?

Good point, I'll remove it for release branches.